CRM/PivotData/DataContact.php

```diff
@@ -0,0 +1,207 @@
+    $fields = array_keys($this->getFields());
+
+    foreach ($fields as $field) {
+      $fieldParts = explode('.', $field);
+      if ($fieldParts[0] === $entityName) {
+        $result[] = $fieldParts[1];
+      }
+    }
+
+    return $result;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  protected function formatResult($data, $dataKey = null, $level = 0) {
+    $result = [];
+    $resultRow = [];
+
+    foreach ($data as $key => $dataRow) {
+      $contactValues = $this->getRowValues($dataRow, 'contact');
+      $contractValues = $dataRow['api.HRJobContract.get']['values'];
+      foreach($contractValues as $contract) {
+        $contractRow = $this->getRowValues($contract, 'jobcontract');
+        $roleRow = $this->getRowValues(array_shift($contract['api.HrJobRoles.get']['values']), 'jobroles');
+        $row = array_merge($contactValues, $contractRow, $roleRow);
+        $row = $this->formatRow($key, $row);
+        $resultRow[] = $row;
+      }
+
+      if(empty($resultRow)) {
+        $resultRow[] = $this->formatRow($key, $contactValues);
+      }
+
+      $headers = array_merge($this->emptyRow, $this->additionalHeaderFields);
+      $finalRow = [];
+      foreach($resultRow as $row) {
+        foreach ($headers as $key => $header) {
+          if (isset($row[$key])) {
+            $finalRow[$key] = $row[$key];
+          }
+          else {
+            $finalRow[$key] = '';
+          }
+
+        }
+        $result[] = $finalRow;
+      }
+
+      $resultRow = [];
+    }
+
+    return $result;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  protected function getEntityIndex(array $row) {
+    return substr($row['Contract Start Date'], 0, 10);
+  }
+
+  /**
+   * @inheritdoc
+   */
+  protected function getFields() {
+    if (empty($this->fields)) {
+      $fields = [];
+      $keys = [];
+      $groups = ['contact', 'jobcontract', 'jobroles'];
+      $result = [];
+
+      // Now get Custom Fields for Contact entity.
+      $customFieldsResult = CRM_Core_DAO::executeQuery(
+        'SELECT g.id AS group_id, f.id AS id, f.label AS label, f.data_type AS data_type, ' .
+        'f.html_type AS html_type, f.date_format AS date_format, og.name AS option_group_name ' .
+        'FROM `civicrm_custom_group` g ' .
+        'LEFT JOIN `civicrm_custom_field` f ON f.custom_group_id = g.id ' .
+        'LEFT JOIN `civicrm_option_group` og ON og.id = f.option_group_id ' .
+        'WHERE g.extends IN(\'Individual\',\'Contact\') AND g.is_active = 1 AND f.is_active = 1 '
+      );
+
+      while ($customFieldsResult->fetch()) {
+        $customField = new CRM_Core_BAO_CustomField();
+        $customField->id = $customFieldsResult->id;
+        $customField->find(true);
+
+        $fields['contact']['custom_' . $customFieldsResult->id] = array(
+          'name' => 'custom_' . $customFieldsResult->id,
+          'title' => $customFieldsResult->label,
+          'pseudoconstant' => array(
+            'optionGroupName' => $customFieldsResult->option_group_name,
```

Comment:
Does every custom field have a pseudoconstant?

Reply:
I copied the code from the implemetation for the Activity report but I checked now and 
not all the custom fields have option_group_id for some of them is null. I can add a check here not to add a pseudoconstant if the optiongroup name is null, but this code catches that when fetching the values for the option group: 
https://github.com/compucorp/uk.co.compucorp.civicrm.pivotreport/blob/a5ea87257dd0bc64c6e394761eb1f7733752cf2b/CRM/PivotData/AbstractData.php#L822